src/main/java/io/debezium/connector/vitess/transaction/OrderedTransactionContext.java

```diff
@@ -0,0 +1,164 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+
+import io.debezium.pipeline.txmetadata.TransactionContext;
+import io.debezium.spi.schema.DataCollectionId;
+
+public class OrderedTransactionContext extends TransactionContext {
+
+    protected static final String OFFSET_TRANSACTION_ID = "transaction_id";
+    protected static final String OFFSET_TRANSACTION_EPOCH = "transaction_epoch";
+    protected static final String OFFSET_TRANSACTION_RANK = "transaction_rank";
+
+    private static final String OFFSET_TABLE_COUNT_PREFIX = "transaction_data_collection_order_";
+    private static final int OFFSET_TABLE_COUNT_PREFIX_LENGTH = OFFSET_TABLE_COUNT_PREFIX.length();
+    private String transactionId = null;
+    private final Map<String, Long> perTableEventCount = new HashMap();
+    private final Map<String, Long> viewPerTableEventCount;
+    private long totalEventCount;
+    private String previousTransactionId = null;
+    private Long transactionEpoch;
+    private BigInteger transactionRank;
+
+    private EpochProvider epochProvider;
+    private RankProvider rankProvider;
+
+    public OrderedTransactionContext(EpochProvider epochProvider, RankProvider rankProvider) {
+        this.epochProvider = epochProvider;
+        this.rankProvider = rankProvider;
+        this.viewPerTableEventCount = Collections.unmodifiableMap(this.perTableEventCount);
+        this.totalEventCount = 0L;
+        this.transactionEpoch = 0L;
+        this.transactionRank = null;
+    }
+
+    private void reset() {
+        this.transactionId = null;
+        this.totalEventCount = 0L;
+        this.perTableEventCount.clear();
+        this.transactionRank = null;
+    }
+
+    @Override
+    public Map<String, Object> store(Map<String, Object> offset) {
+        if (!Objects.isNull(this.transactionId)) {
+            offset.put(OFFSET_TRANSACTION_ID, this.transactionId);
+        }
+        if (!Objects.isNull(this.transactionEpoch)) {
+            offset.put(OFFSET_TRANSACTION_EPOCH, this.transactionEpoch);
+        }
+        if (!Objects.isNull(this.transactionRank)) {
+            offset.put(OFFSET_TRANSACTION_RANK, this.transactionRank.toString());
+        }
+
+        Iterator var3 = this.perTableEventCount.entrySet().iterator();
+
+        while (var3.hasNext()) {
+            Map.Entry<String, Long> e = (Map.Entry) var3.next();
+            offset.put(OFFSET_TABLE_COUNT_PREFIX + e.getKey(), e.getValue());
+        }
+        return offset;
+    }
+
+    public static OrderedTransactionContext load(Map<String, ?> offsets, EpochProvider epochProvider, RankProvider rankProvider) {
+        OrderedTransactionContext context = new OrderedTransactionContext(epochProvider, rankProvider);
+        context.transactionId = (String) offsets.get(OFFSET_TRANSACTION_ID);
+        context.previousTransactionId = (String) offsets.get(OFFSET_TRANSACTION_ID);
```

Comment:
`transactionId` and `previousTransactionId` is using the same key in the `offsets` map?

Reply:
The offset_transaction_id stored in offset map is for a specific shard.  A VGTID in the same offset map represents multiple shards.  If the offset_transaction_id in the offset map represents shard1, but the new events coming from VSTREAM is for shard2, how do you do the comparison?

Reply:
Ah yes forgot to add this part, we do need mapping of shard to epoch.